### DISCUSSION TITLE
Optimising middleware

### DISCUSSION THREAD

**Discussion by wiebekaai** [vercel/next.js#47106] [Help]
### Summary

Middleware is being called on every route, even the prefetches for `Link`s. I'm only using middleware to get the users geolocation - basically needed only once per user, but now I get millions of invocations all showing up in our usage.

Is this being worked on or does anyone have a solution?

I thought middleware was a perfect fit, but maybe I'm better off having a serverless/edge function and call it on the client side.

Thanks.

**Comment by petmarsh** [comment]
That's the purpose of middleware ;)
From the docs: Middleware will be invoked for **every route in your project**.

maybe move it to e.g. root layout or some location provider component you'd wrap the page into, memoize the computed value, ...

**Reply by wiebekaai** [reply]
Thanks for the input. Yea, client side is the way to go. Not being able to have more targeted middleware is a big flaw imo. There are very little use cases to apply middleware to every route.

**Comment by icyJoseph** [comment]
Use, a matcher https://nextjs.org/docs/advanced-features/middleware#matcher, so that you only activate the middleware for actual routes that your users consume directly.

Otherwise, internals, and a bunch of other assets end up activating the middleware.
